### ISSUE TITLE
use server  exec-mode,this error:the plugin you are using seems to generated a bad URL.

### ISSUE THREAD

**Issue by iT-Boyer** [skuro/plantuml-mode#146]
# Summary
![截屏2021-07-10 上午10 04 37](https://user-images.githubusercontent.com/4022953/125148608-3eb69900-e166-11eb-9411-8fc0ab47caad.png)

> Write here the summary of your changes

# Checklist for pull requests

- [ x ] I'm fine with my code being released under the GPLv2 license
- [ x ] the pull request is filed against the `develop` branch instead of `master`

**Comment by kasimirmoilanen** [comment]
[Here](https://forum.plantuml.net/11432/the-plugin-you-are-using-seems-generate-private-gitlab-server) the PlantUML team gives a more detailed explanation why this happens.

**Comment by natrys** [comment]
@kasimirmoilanen do you find that prefixing data with `~1` header as suggested above, fixes the problem?

I just tried that (by patching line 432 in plantuml-mode.el), but I still get error saying:

> This URL does not look like HUFFMAN data

**Comment by kasimirmoilanen** [comment]
@natrys I tried this as a quick fix, but I also still get the error. I just changed plantuml-default-exec-mode to jar as specified in the readme, I think that or executable mode are better anyway as you don't need to connect to external servers.

**Comment by natrys** [comment]
@kasimirmoilanen  Yeah you are right. I was looking to avoid java things in all my machines, but I suppose it is what it is.

Anyway, looking at the elisp function, it appears that the function was merely doing base64 encoding, without performing the compression (hoffman or deflate) required before that, so no wonder it doesn't work.

As for fixing the issue, Emacs sadly doesn't have compression primitives like zlib or brotli, which appears to be the server's recommendation now. Emacs actually builds with zlib, but the FFI only exposes decompression. I found that someone sent patch to add compression too in mailing list, but somewhat true to form it got lost in discussion without actually getting merged:

https://lists.gnu.org/r/emacs-devel/2020-03/msg00802.html

We could just shell out to gzip I guess. But as I try this, it appears that the bigger problem is that the server seems to expect a modified base64 encoding for legacy reasons:

https://plantuml.com/en/text-encoding

So off the shelf base64 from coreutils or `base64-encode-string` shouldn't have been able to work anyway.

That page also lists that now API also accepts hex encoding, which is finally something that's doable in pure elisp. This of course increases size 2x from plaintext, which is kind of terrible. But for now I don't really care:

```elisp
(defun hex-encode (str)
  (string-join (mapcar (lambda (c) (format "%02x" c)) str)))

(defun plantuml-server-encode-url (string)
  "Encode the string STRING into a URL suitable for PlantUML server interactions."
  (let* ((encoded-string (hex-encode string)))
    (concat plantuml-server-url "/" plantuml-output-type "/~h" encoded-string)))
```

And this worked. 

(well the server `https` endpoint appears to be borked so I had to change server url to http, which is again quite terrible, but whatever)

**Comment by jimmy-wong-beamery** [comment]
Server mode no longer works out of the box because of this issue, but jar mode works tho.

**Comment by jjnilton** [comment]
@natrys method does work for me (thanks for sharing), but special/accented characters are not rendered correctly:

![image](https://user-images.githubusercontent.com/228200/205492788-145cfe63-0c24-4893-8521-bc7a7ed726ac.png)

Does anyone know a workaround for that?

**Comment by natrys** [comment]
@jjnilton yeah I failed to account for multi byte code points. I am not really knowledgeable around these APIs in Emacs, but this might solve your issue:

``` elisp
(defun hex-encode (str)
  (string-join (mapcar (lambda (c) (format "%02x" c)) (string-as-unibyte str))))
```

On a tangential note, I had discovered https://kroki.io/ who provide a unified API for quite a number of backends including plantuml. I have brushed up my old wrapper code that shows preview using it, in case it's useful to anyone: [diagram-preview](https://github.com/natrys/diagram-preview).

**Comment by jjnilton** [comment]
That solved it, thanks @natrys. Good to know about kroki.io, looks great, thanks for sharing your package.

**Comment by francesco-cadei** [comment]
Thanks @natrys !
I set to:
```
(use-package plantuml-mode
  :config
  (eval-and-compile (defun hex-encode (str)
		      (string-join (mapcar (lambda (c) (format "%02x" c)) (encode-coding-string str 'utf-8)))))
  (defun plantuml-server-encode-url (string)
    "Encode the string STRING into a URL suitable for PlantUML server interactions."
    (let* ((encoded-string (hex-encode string)))
      (concat plantuml-server-url "/" plantuml-output-type "/~h" encoded-string)))
  :mode ("\\.puml\\'" . plantuml-mode))
;; M-x plantuml-download-jar
```

**Comment by skuro** [comment]
This issue should be now fixed in the new versions of `plantuml-mode` (it now uses HEX encoding indeed)
